### DISCUSSION TITLE
Extension `.langium-types` for files containing only declared types

### DISCUSSION THREAD

**Discussion by pluralia** [eclipse-langium/langium#757] [Ideas]
There can be many `.langium` files in the workspace -- some of them contain only parser rules and some contain only types, so it makes sense to add the `.langium-types` extension to the files that contain only declared types. This is especially important after the [PR about generating declared types](https://github.com/langium/langium/pull/754) merge, where you can get declared types automatically in a separate file.

**Comment by msujew** [comment]
What is the actual benefit behind this? I would get the separation of concerns if we were to disable type declarations in `.langium` files, but that would basically just create a separate language with a subset of an existing language. Seems to me like a lot of effort without any advantages for anyone.

**Reply by pluralia** [reply]
When your project is mature, it's nice to put types separately since you don't want to change them often. With `.langium-types` you can easily see which files are about parser rules and which are about types. It's like `hpp` and `cpp` files.

If you use declared types generation, generated types will be in another file, and when you import this file in your initial grammar, you can't duplicate definitions of these types. As a result, we get 2 files related to 1 grammar: 1 file has parser rules and 1 file has their types.

**Reply by msujew** [reply]
>  It's like hpp and cpp files.

I wouldn't say that cpp/c is the pinnacle of a well designed language. If anything, that's more of a counter argument...

> When your project is mature, it's nice to put types separately since you don't want to change them often.

Sure, but what is the advantage of building an entire subset language for this? We already have a language which is perfectly capable of allowing you to split your grammar/type definitions into separate files, while also allowing to have them in a single file.

**Reply by spoenemann** [reply]
I'd look at TypeScript as example rather than C++. If you'd want to separate your language declaration into syntax and types, you'd create files like `mylang-syntax.ts` and `mylang-types.ts` in TypeScript. Similarly, you can create files `mylang-syntax.langium` and `mylang-types.langium`.

**Reply by pluralia** [reply]
> Similarly, you can create files mylang-syntax.langium and mylang-types.langium.

Are you suggesting that we simply introduce a naming convention for such cases?

**Reply by spoenemann** [reply]
Yes, but we shouldn't enforce it. Every user is free to create their own naming convention. We should try to apply consistent naming to our own projects.
